File: php/src/Support/PseudoId.php
<?php

declare(strict_types=1);

namespace App\Support;

/**
 * There is no `users` collection in this app's data model (see plan/build-plan.md), so an
 * assignee is captured as a free-typed name rather than looked up from a directory. Verified live
 * against the real project (see the reference web app's build-plan.md "Live smoke test results"):
 * `cards.assigneeId` is schema-typed as an ObjectId reference, not a free-form string as
 * originally assumed — a plain slug like "mia-member" is rejected with "Invalid ObjectId format
 * for assigneeId". This derives a stable, valid-looking 24-hex-char id from the typed name (the
 * same name always maps to the same id) purely so `assigneeId` passes that format check and is
 * never left orphaned from `assigneeName` — it is not a real user lookup key. A PHP port of the
 * reference app's `pseudoObjectId()` (src/lib/utils.ts), using the same djb2 string hash.
 */
final class PseudoId
{
    private const UINT32_MODULUS = 4294967296; // 2^32
    private const INT32_MAX = 2147483648; // 2^31

    public static function pseudoObjectId(string $seed): string
    {
        $trimmed = strtolower(trim($seed));
        $hex = '';
        $round = 0;
        while (strlen($hex) < 24) {
            $hex .= str_pad(dechex(self::djb2("{$trimmed}:{$round}")), 8, '0', STR_PAD_LEFT);
            $round++;
        }
        return substr($hex, 0, 24);
    }

    /** Mirrors the reference app's `djb2()`, including its JS Int32 bitwise-op wraparound. */
    private static function djb2(string $str): int
    {
        $hash = 5381;
        $length = strlen($str);
        for ($i = 0; $i < $length; $i++) {
            $hash = self::toInt32($hash * 33) ^ ord($str[$i]);
        }
        return $hash < 0 ? $hash + self::UINT32_MODULUS : $hash;
    }

    /** Wraps an arbitrary integer into JS's signed Int32 range, the same way `x | 0` / bitwise ops would. */
    private static function toInt32(int $n): int
    {
        $n %= self::UINT32_MODULUS;
        if ($n >= self::INT32_MAX) {
            $n -= self::UINT32_MODULUS;
        } elseif ($n < -self::INT32_MAX) {
            $n += self::UINT32_MODULUS;
        }
        return $n;
    }
}

File: go/internal/store/pseudoid.go
package store

import (
	"fmt"
	"strings"
)

// djb2 is a small deterministic string hash - used below to derive a stable pseudo-ObjectId,
// nothing more. Ported bit-for-bit from the reference web app's src/lib/utils.ts.
func djb2(str string) uint32 {
	var hash uint32 = 5381
	for _, r := range str {
		hash = (hash * 33) ^ uint32(r)
	}
	return hash
}

// pseudoObjectID derives a stable, valid-looking 24-hex-char id from a typed assignee name, purely
// so `cards.assigneeId` passes Mudbase's ObjectId format check and is never left orphaned from
// `assigneeName`. There is no `users` collection in this app's data model (see
// plan/build-plan.md), so an assignee is captured as a free-typed name rather than looked up from
// a directory - but the platform's query sanitizer validates any field named `...Id` as an
// ObjectId reference (confirmed live: a plain slug like "mia-member" is rejected with "Invalid
// ObjectId format for assigneeId"). The same name always maps to the same id; this is not a real
// user lookup key. Ported bit-for-bit from the reference web app's src/lib/utils.ts
// (pseudoObjectId).
func pseudoObjectID(seed string) string {
	trimmed := strings.ToLower(strings.TrimSpace(seed))
	var hex strings.Builder
	for round := 0; hex.Len() < 24; round++ {
		hex.WriteString(fmt.Sprintf("%08x", djb2(fmt.Sprintf("%s:%d", trimmed, round))))
	}
	return hex.String()[:24]
}
